lintreview/tools/goodcheck.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import absolute_import
+import logging
+import json
+import lintreview.docker as docker
+from lintreview.tools import Tool
+
+log = logging.getLogger(__name__)
+
+
+class Goodcheck(Tool):
+
+    name = 'goodcheck'
+
+    def check_dependencies(self):
+        """
+        See if ruby image exists
+        """
+        return docker.image_exists('ruby2')
+
+    def process_files(self, files):
+        """
+        Run checks with goodcheck
+        """
+        log.debug('Processing %s files with %s', files, self.name)
+        command = self._create_command()
+        command += files
+        output = docker.run('ruby2', command, self.base_path)
+
+        if not output:
+            log.debug('No goodcheck errors found.')
+            return False
+
+        output = output.split("\n")
+        self._process_output(output, docker.strip_base)
+
+    def _create_command(self):
+        command = ['goodcheck', 'check', '--format', 'json']
+        if self.options.get('rules'):
+            for rule in self.options['rules'].split(','):
+                command.extend(['-R', rule.strip()])
+        if self.options.get('config'):
+            command.extend(['--config', self.options.get('config')])
+        return command
+
+    def _process_output(self, output, filename_converter):
+        """
+        Process goodcheck json results.
+
+        Parses a line from `output` containing check results, formatted like:
+            [{"rule_id":"<id>","path":"<filename>",
+              "location":{"start_line":<line>,"start_column":<col>,
+                          "end_line":<endline>,"end_column":<endcol>},
+              "message":"<message>",
+              "justifications":[]}]
+        """
+        results = []
+        # When outputting json, goodcheck prints messages to stderr about which
+        # files it's checking. IF there is any data it will probably be in the
+        # second-to-last line
+        for line in reversed(output):
```

Comment:
Is there one JSON object for each file with problems or one for all problems?

Reply:
I played with the goodcheck tool a little just now. With every combination of flags and arguments I can think to try, the tool prints a JSON document, on one line, containing all problems, to stdout.

Looping over the tool output looking for something to unserialize using json.loads is admittedly a little hacky. The better fix might be to add an option to change how individual tools deal with commands that emit output to stderr, but it seemed like the other tools implement their own logic as needed to deal with non-standard output.

Reply:
Separating steer and strout might be a good idea. There are a few tools that have workarounds for the simplistic output handling in place now.